compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -391,47 +600,68 @@ runEnv comp env0 = do
   return env1
 
 -- | Skolemise an expression variable and extend the environment.
-extVarEnv :: MonadEnv m ph
+extVarEnv :: (IsPhase ph, MonadEnv m ph)
   => ExprVarName
   -- ^ The expression variable to be skolemised.
   -> m ()
 extVarEnv x = extSkolEnv (SkolVar x)
 
 -- | Skolemise a list of record projection and extend the environment.
-extRecEnv :: MonadEnv m ph
+extRecEnv :: (IsPhase ph, MonadEnv m ph)
   => ExprVarName
   -- ^ The variable on which is being projected.
   -> [FieldName]
   -- ^ The fields which should be skolemised.
   -> m ()
-extRecEnv x fs = extSkolEnv (SkolRec x fs)
+extRecEnv x fs = do
+  env <- getEnv
+  let skols = envSkols env
+      curFs = [fs' | SkolRec x' fs' <- skols, x == x']
+      newFs = if null curFs
+        then fs
+        else fs ++ head curFs
+  extSkolEnv (SkolRec x newFs)
+
+-- | Extend the environment with the fields of any given record or type
+-- constructor type.
+extRecEnvTCons :: (IsPhase ph, MonadEnv m ph)
+  => [(FieldName, Type)]
+  -- ^ The given fields and their corresponding types to analyse.
+  -> m ()
+extRecEnvTCons = mapM_ step
+  where
+    step :: (IsPhase ph, MonadEnv m ph) => (FieldName, Type) -> m ()
+    step (f,typ) =
+      recTypFields typ >>= \case
+        Nothing -> return ()
+        Just fsRec -> do
+          extRecEnv (fieldName2VarName f) $ map fst fsRec
```

Comment:
```suggestion
    step (f,typ) = do
      mbFields <- recTypFields typ
      whenJust mbFields $ \fsRec -> 
        extRecEnv (fieldName2VarName f) $ map fst fsRec
```
A bit easier to parse. Side note: Do you ever expect `recTypFields` to return `Nothing` here? It looks like you are only calling this in cases where the type must be a record so maybe crashing is more appropriate?